File: prefix_sum/2968_hard/max_frequency_score.py

def maximize_frequency_score(nums: list[int], limit: int) -> int:  # LeetCode Q.2968.
    nums.sort()
    prefix_sums = []
    for num in nums:
        if not prefix_sums:
            prefix_sums.append(num)
            continue
        prefix_sums.append(prefix_sums[-1] + num)

    max_score = 1  # Base case.

    left_idx, mid_idx = 0, 0
    for right_idx in range(len(nums)):
        first_while = True
        while True:
            if not first_while:
                left_idx += 1  # Since 2nd while, the subarray left bound rises.

            # Always optimal to change all nums to the num at mid idx.
            mid_idx = (left_idx + right_idx) // 2
            operations = prefix_sums[right_idx] - prefix_sums[mid_idx]
            operations -= nums[mid_idx] * (right_idx - mid_idx)

            operations += nums[mid_idx] * (mid_idx - left_idx)
            if mid_idx > 0:
                operations -= prefix_sums[mid_idx - 1]
                if left_idx > 0:
                    operations += prefix_sums[left_idx - 1]

            if limit >= operations or left_idx == right_idx:
                break
            first_while = False

        score = right_idx + 1 - left_idx
        if score > max_score:
            max_score = score

    return max_score

File: prefix_sum/2968_hard/max_frequency_score.cpp
#include <vector>
#include <algorithm>
using namespace std;

int maximizeFrequencyScore(vector<int> &nums, long long limit)
{ // LeetCode Q.2968.
    sort(nums.begin(), nums.end());
    vector<long long> prefixSums;

    for (auto num : nums)
    {
        if (prefixSums.empty())
        {
            prefixSums.push_back(num);
            continue;
        }
        prefixSums.push_back(prefixSums.back() + num);
    }

    int maxScore = 1; // Base case.
    long long operations = 0;

    long long leftIdx = 0, midIdx = 0; // Set long long to prevent overflow.

    for (long long rightIdx = 0; rightIdx < nums.size(); rightIdx++)
    {
        bool firstWhile = true;
        do
        {
            if (firstWhile == false)
                leftIdx++; // Since 2nd while, the subarray left bound rises.

            // Always optimal to change all nums to the num at mid idx.
            midIdx = (leftIdx + rightIdx) / 2;

            operations = prefixSums[rightIdx] - prefixSums[midIdx];
            operations -= nums[midIdx] * (rightIdx - midIdx);

            operations += nums[midIdx] * (midIdx - leftIdx);

            if (midIdx > 0)
                operations -= prefixSums[midIdx - 1];

            if (leftIdx > 0)
                operations += prefixSums[leftIdx - 1];

            firstWhile = false;
        } while (limit < operations && leftIdx < rightIdx);

        int score = rightIdx + 1 - leftIdx;
        if (score > maxScore)
            maxScore = score;
    }

    return maxScore;
}
